File: 2019/territoriali/antivirus/solutions/sol_from_stub.c
#include <stdio.h>
#include <stdbool.h>

#define MAXN 1000

bool compare(char* s1, char* s2, int c) {
    while (c--) {
        if (*s1 != *s2) return false;
        s1++;
        s2++;
    }
    return true;
}

void solve(int t) {
    int N1, N2, N3, N4;
    int M;
    char F1[MAXN+1], F2[MAXN+1], F3[MAXN+1], F4[MAXN+1];

    // scrivi in queste variabili la risposta
    int p1, p2, p3, p4;

    scanf("%d %d %d %d", &N1, &N2, &N3, &N4);
    scanf("%d", &M);
    scanf("%s %s %s %s", F1, F2, F3, F4);

    for (int i = 0; i <= N1-M; i++) {
        p1 = i;
        p2 = p3 = p4 = -1;
        for (int j = 0; j <= N2-M && p2 < 0; j++)
            if (compare(F1+i, F2+j, M))
                p2 = j;
        for (int j = 0; j <= N3-M && p3 < 0; j++)
            if (compare(F1+i, F3+j, M))
                p3 = j;
        for (int j = 0; j <= N4-M && p4 < 0; j++)
            if (compare(F1+i, F4+j, M))
                p4 = j;
        if (p2 != -1 && p3 != -1 && p4 != -1)
            break;
    }

    printf("Case #%d: %d %d %d %d\n", t, p1, p2, p3, p4);
}

int main() {
    // le seguenti due righe possono essere utili per leggere e scrivere da file

    //freopen("input.txt", "r", stdin);
    //freopen("output.txt", "w", stdout);

    int T, t;
    scanf("%d", &T);

    for (t = 1; t <= T; t++) {
        solve(t);
    }
}

File: 2019/territoriali/antivirus/solutions/sol_from_stub.pas
const MAXN = 1000;

procedure solve(test: longint);
var
    N1, N2, N3, N4: longint;
    M: longint;
    F1, F2, F3, F4: ansistring;

    (* scrivi in queste variabili la risposta *)
    p1, p2, p3, p4: longint;

    i: longint;
    ok: boolean;

begin

    readln(N1, N2, N3, N4);
    readln(M);
    readln(F1);
    readln(F2);
    readln(F3);
    readln(F4);

    for p1 := 0 to (N1 - M) do
        for p2 := 0 to (N2 - M) do
            for p3 := 0 to (N3 - M) do
                for p4 := 0 to (N4 - M) do
                begin
                    ok := True;
                    for i := 1 to M do
                        if (F1[p1 + i] <> F2[p2 + i]) or (F2[p2 + i] <> F3[p3 + i]) or (F3[p3 + i] <> F4[p4 + i]) then
                            ok := False;

                    if ok then
                        writeln('Case #', test, ': ', p1, ' ', p2, ' ', p3, ' ', p4);
                end;
end;


var
    T, test: longint;

begin
    (* le seguenti due righe possono essere utili per leggere e scrivere da file *)

    (*
    assign(input, 'input.txt'); reset(input);
    assign(output, 'output.txt'); rewrite(output);
    *)

    readln(T);

    for test := 1 to T do
        solve(test);
end.
